File: app/lib/config/crn.ts
/**
 * CRN (Client Reference Number) configuration + helpers.
 *
 * The CRN is the canonical identifier for an external client. How the value is
 * sourced is configured in `app.config.ts` (committed, non-secret) under `crn`:
 *   - autoGenerate=false (default) — users type an existing external CRN,
 *     validated for format + uniqueness.
 *   - autoGenerate=true            — the system generates the next CRN when a new
 *     client is registered, formatted as `${prefix}${zero-padded counter}`.
 */
import type { Tx } from '../db';
import { appConfig } from '../../../app.config';

export interface CrnConfig {
  autoGenerate: boolean;
  prefix: string;
  pad: number;
  /** Allowed shape for a manually-entered CRN (already normalized to uppercase). */
  pattern: RegExp;
}

export function crnConfig(): CrnConfig {
  const { autoGenerate, prefix, pad } = appConfig.crn;
  return {
    autoGenerate,
    prefix: prefix ?? 'CRN-',
    pad: Number.isFinite(pad) && pad > 0 ? pad : 6,
    // 3–32 chars: uppercase alphanumerics and dashes, must start alphanumeric.
    pattern: /^[A-Z0-9][A-Z0-9-]{2,31}$/,
  };
}

/** Trim and uppercase so CRNs compare case-insensitively against the PK. */
export function normalizeCrn(raw: string): string {
  return raw.trim().toUpperCase();
}

/**
 * Placeholder CRN prefix. A client can be registered before its real CRN is known;
 * the system assigns it a `PENDING-000001`-style CRN and flags `crn_pending` so the
 * UI can highlight it and prompt for the real value later.
 */
export const PENDING_CRN_PREFIX = 'PENDING-';

/** True when `crn` is a system-generated placeholder awaiting a real value. */
export function isPendingCrn(crn: string): boolean {
  return crn.toUpperCase().startsWith(PENDING_CRN_PREFIX);
}

/**
 * Reserve and return the next placeholder CRN (`PENDING-000001`, …). Like
 * generateNextCrn, MUST run inside an executeTransaction callback so the max-scan
 * and insert are atomic. Loops past the rare case where the candidate already exists.
 */
export function generatePendingCrn(tx: Tx): string {
  const pad = 6;
  for (;;) {
    const row = tx.get<{ maxn: number | null }>(
      `SELECT MAX(CAST(substr(crn, ?) AS INTEGER)) AS maxn FROM clients WHERE crn LIKE ?`,
      [PENDING_CRN_PREFIX.length + 1, `${PENDING_CRN_PREFIX}%`]
    );
    const n = (row?.maxn ?? 0) + 1;
    const candidate = `${PENDING_CRN_PREFIX}${String(n).padStart(pad, '0')}`;
    if (!tx.get(`SELECT 1 FROM clients WHERE crn = ?`, [candidate])) return candidate;
  }
}

/** True when `crn` (already normalized) is a syntactically valid CRN. */
export function isValidCrn(crn: string): boolean {
  return crnConfig().pattern.test(crn);
}

/**
 * Reserve and return the next auto-generated CRN. MUST be called inside an
 * executeTransaction callback — better-sqlite3 is synchronous and serialized, so
 * the read-increment within the transaction is atomic (no counter race). Loops
 * past the rare case where the formatted candidate collides with a CRN that was
 * entered manually in the same shape.
 */
export function generateNextCrn(tx: Tx): string {
  const { prefix, pad } = crnConfig();
  for (;;) {
    const row = tx.get<{ next_value: number }>(`SELECT next_value FROM crn_sequence WHERE id = 1`);
    const n = row?.next_value ?? 1;
    tx.run(`UPDATE crn_sequence SET next_value = ? WHERE id = 1`, [n + 1]);
    const candidate = `${prefix}${String(n).padStart(pad, '0')}`.toUpperCase();
    const clash = tx.get(`SELECT 1 FROM clients WHERE crn = ?`, [candidate]);
    if (!clash) return candidate;
  }
}

File: backend/crm_sync/db/crn.py
"""
CRN (Client Reference Number) helpers — a line-for-line port of app/lib/config/crn.ts.

The CRN is the canonical identity of an external client. Two ways it gets assigned:

  * autoGenerate = False (the current setting) — a human types the client's real CRN from
    whatever upstream system owns it. If they don't know it yet, the client can be registered
    with a `PENDING-000001` placeholder and flagged so the UI nags for the real value later.
  * autoGenerate = True — the app hands out the next CRN from a counter table.

Both generators MUST be called from inside an open `BEGIN IMMEDIATE` transaction. The
TypeScript versions carry the same requirement (they take a `Tx` handle and say so): the
read-then-insert is only atomic because the write lock is already held. Calling them in
autocommit mode would let two processes reserve the same CRN.

Keep this file in sync with crn.ts by hand.
"""

import sqlite3

from ..config import CRN_PATTERN, PENDING_CRN_PREFIX, TABLE_CLIENTS, TABLE_CRN_SEQUENCE, CrmConfig


def normalize_crn(raw: str) -> str:
    """
    Trim and uppercase, so CRNs compare case-insensitively against the primary key.

    `clients.crn` is a plain TEXT PRIMARY KEY (case-sensitive), and the app only ever stores
    normalized values — so normalizing on the way in is what makes 'crn-42' find 'CRN-42'.
    """
    return raw.strip().upper()


def is_valid_crn(crn: str) -> bool:
    """True when an already-normalized CRN matches the configured shape."""
    return bool(CRN_PATTERN.match(crn))


def is_pending_crn(crn: str) -> bool:
    """True when `crn` is a system-generated placeholder still awaiting its real value."""
    return crn.upper().startswith(PENDING_CRN_PREFIX)


def generate_pending_crn(cur: sqlite3.Cursor) -> str:
    """
    Reserve and return the next placeholder CRN: PENDING-000001, PENDING-000002, ...

    Must run inside `BEGIN IMMEDIATE`. Scans for the highest existing placeholder number and
    adds one, then loops in the (vanishing) case where the formatted candidate already exists.

    Mirrors generatePendingCrn() in crn.ts, including the `substr(crn, len(prefix) + 1)`
    offset — SQLite's substr is 1-indexed.
    """
    pad = 6
    while True:
        row = cur.execute(
            f"SELECT MAX(CAST(substr(crn, ?) AS INTEGER)) AS maxn FROM {TABLE_CLIENTS} WHERE crn LIKE ?",
            (len(PENDING_CRN_PREFIX) + 1, f"{PENDING_CRN_PREFIX}%"),
        ).fetchone()
        n = (row["maxn"] if row and row["maxn"] is not None else 0) + 1
        candidate = f"{PENDING_CRN_PREFIX}{n:0{pad}d}"
        exists = cur.execute(f"SELECT 1 FROM {TABLE_CLIENTS} WHERE crn = ?", (candidate,)).fetchone()
        if not exists:
            return candidate


def generate_next_crn(cur: sqlite3.Cursor, cfg: CrmConfig) -> str:
    """
    Reserve and return the next auto-generated CRN from the `crn_sequence` counter.

    Must run inside `BEGIN IMMEDIATE`. Reads and bumps the counter under the write lock, then
    loops past the rare case where the formatted candidate collides with a CRN that was typed
    in manually in the same shape.

    Only reachable when `cfg.crn.auto_generate` is True. Mirrors generateNextCrn() in crn.ts.
    """
    prefix, pad = cfg.crn.prefix, cfg.crn.pad
    while True:
        row = cur.execute(f"SELECT next_value FROM {TABLE_CRN_SEQUENCE} WHERE id = 1").fetchone()
        n = row["next_value"] if row else 1
        cur.execute(f"UPDATE {TABLE_CRN_SEQUENCE} SET next_value = ? WHERE id = 1", (n + 1,))
        candidate = f"{prefix}{n:0{pad}d}".upper()
        clash = cur.execute(f"SELECT 1 FROM {TABLE_CLIENTS} WHERE crn = ?", (candidate,)).fetchone()
        if not clash:
            return candidate
